<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Nucleotide_Essentials.jl · Nucleotide_Essentials</title><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>Nucleotide_Essentials</a></span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Nucleotide_Essentials.jl</a><ul class="internal"><li><a class="tocitem" href="#Data-Types"><span>Data Types</span></a></li><li><a class="tocitem" href="#Functions"><span>Functions</span></a></li><li><a class="tocitem" href="#Index"><span>Index</span></a></li><li><a class="tocitem" href="#Change-Log"><span>Change Log</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Nucleotide_Essentials.jl</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Nucleotide_Essentials.jl</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/phorve/Nucleotide_Essentials.jl/blob/master/docs/src/index.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Nucleotide_Essentials.jl"><a class="docs-heading-anchor" href="#Nucleotide_Essentials.jl">Nucleotide_Essentials.jl</a><a id="Nucleotide_Essentials.jl-1"></a><a class="docs-heading-anchor-permalink" href="#Nucleotide_Essentials.jl" title="Permalink"></a></h1><ul><li><a href="#Nucleotide_Essentials.jl">Nucleotide_Essentials.jl</a></li><li class="no-marker"><ul><li><a href="#Data-Types">Data Types</a></li><li><a href="#Functions">Functions</a></li><li><a href="#Index">Index</a></li><li><a href="#Change-Log">Change Log</a></li></ul></li></ul><h2 id="Data-Types"><a class="docs-heading-anchor" href="#Data-Types">Data Types</a><a id="Data-Types-1"></a><a class="docs-heading-anchor-permalink" href="#Data-Types" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="Nucleotide_Essentials.FastqRecord" href="#Nucleotide_Essentials.FastqRecord"><code>Nucleotide_Essentials.FastqRecord</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">Nucleotide_Essentials.FastqRecord</code></pre><p><strong>Components</strong></p><ul><li>ID: The unique sequence identifier associated with that entry</li><li>sequence: The nucleotide sequence of that entry</li><li>quality: The quality scores of that entry </li><li>filename: The original file name </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/phorve/Nucleotide_Essentials.jl/blob/ef9c5233eb534080761fa3a827c431069ffb4fe6/src/Reading/readFastq.jl#L1-L8">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Nucleotide_Essentials.FastaRecord" href="#Nucleotide_Essentials.FastaRecord"><code>Nucleotide_Essentials.FastaRecord</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">Nucleotide_Essentials.FastaRecord</code></pre><p><strong>Components</strong></p><ul><li>ID: The unique sequence identifier associated with that entry</li><li>sequence: The nucleotide sequence of that entry</li><li>filename: The original file name </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/phorve/Nucleotide_Essentials.jl/blob/ef9c5233eb534080761fa3a827c431069ffb4fe6/src/Reading/readFasta.jl#L1-L9">source</a></section></article><h2 id="Functions"><a class="docs-heading-anchor" href="#Functions">Functions</a><a id="Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Functions" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="Nucleotide_Essentials.readFastq" href="#Nucleotide_Essentials.readFastq"><code>Nucleotide_Essentials.readFastq</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">Nucleotide_Essentials.readFastq
readFastq(Path::String)</code></pre><p>.fastq file =&gt; readFastq(Path) =&gt; FastqRecord(ID, sequence, quality, filename)</p><p>supported keyword arguments include: </p><ul><li><code>Path::String</code>: The full or relative path to a .fastq file</li></ul><p><strong>Example:</strong></p><pre><code class="language-julia hljs"># Supply the path to a .Fastq file that you would like to import
myfastq = readFastq(&quot;myfastq.fastq&quot;)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/phorve/Nucleotide_Essentials.jl/blob/ef9c5233eb534080761fa3a827c431069ffb4fe6/src/Reading/readFastq.jl#L16-L30">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Nucleotide_Essentials.readFasta" href="#Nucleotide_Essentials.readFasta"><code>Nucleotide_Essentials.readFasta</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">Nucleotide_Essentials.readFasta</code></pre><p>Imports a .fasta file into julia</p><pre><code class="nohighlight hljs">readFasta(Path::String)</code></pre><p>.fasta file =&gt; readFasta(Path) =&gt; FastaRecord(ID, sequence, filename)</p><p>Supported keyword arguments include: </p><ul><li><code>Path::String</code>: The full or relative path to a .fasta file</li></ul><p><strong>Example:</strong></p><pre><code class="language-julia hljs"># Supply the path to a .fasta file that you would like to import - it is recommended to include `;` in your command to prevent printing potentially large .fasta files in the REPL
myfasta = readFasta(&quot;myfasta.fasta&quot;);</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/phorve/Nucleotide_Essentials.jl/blob/ef9c5233eb534080761fa3a827c431069ffb4fe6/src/Reading/readFasta.jl#L16-L34">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Nucleotide_Essentials.writeFasta" href="#Nucleotide_Essentials.writeFasta"><code>Nucleotide_Essentials.writeFasta</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">Nucleotide_Essentials.writeFasta</code></pre><p>readFasta(Path::String) FastaRecord =&gt; write<em>fasta(input</em>fasta, out, compressed) =&gt; .fasta file/.fasta.gz file </p><p>Creates a single or multiple entry FastaRecord and outputs either a .fasta or compressed .fasta.gz file to the desired directory</p><p>Supported keyword arguments include: </p><ul><li><code>input_fasta::FastaRecord</code>: A FastaRecord with either a single entry or multiple entries </li><li><code>out::String</code>: The full or relative path to the directory where files should be written to</li><li><code>compressed::Bool</code>: Whether or not to write the .fasta files as compressed files or not. <ul><li>If <code>true</code>, files will written as .fasta.gz files</li><li>If <code>false</code>, files will written as .fasta files</li></ul></li></ul><p><strong>Example:</strong></p><pre><code class="language-julia hljs"># .fasta files can be written as from an already imported FastaRecord in Julia 
myfasta = readFasta(&quot;myfasta.fasta&quot;);
writeFasta(input_fasta, &quot;example/output/directory, false)

# .fasta files can be written as a .fasta.gz from an already imported FastaRecord in Julia 
myfasta = readFasta(&quot;myfasta.fasta&quot;);
writeFasta(input_fasta, &quot;example/output/directory&quot;, true)

# .fasta files with multiple sequences can be read and written as individual .fasta or .fasta.gz in the same step
myfasta = readFasta(&quot;myfasta.fasta&quot;);
writeFasta(readFasta(&quot;/myfasta.fasta&quot;), &quot;example/output/directory&quot;, true);</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/phorve/Nucleotide_Essentials.jl/blob/ef9c5233eb534080761fa3a827c431069ffb4fe6/src/Writing/writeFasta.jl#L1-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Nucleotide_Essentials.FastqtoFasta" href="#Nucleotide_Essentials.FastqtoFasta"><code>Nucleotide_Essentials.FastqtoFasta</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">Nucleotide_Essentials.FastqtoFasta</code></pre><p>Converts a FastqRecord to a FastaRecord. Can also input and convert a .fastq file to a FastaRecord in the same function. </p><p>FastqtoFasta(Fastq::Union{String, FastqRecord}) .fastq file =&gt; FastqtoFasta(Fastq) =&gt; FastaRecord(ID, sequence, filename) FastqRecord(ID, sequence, quality, filename) =&gt; FastqtoFasta(Fastq) =&gt; FastaRecord(ID, sequence, filename)</p><p>Supported keyword arguments include: </p><ul><li><code>Fastq::Union{String, FastqRecord}</code>: <ul><li>The full or relative path to a .fastq file</li><li>A FastqRecord</li></ul></li></ul><p><strong>Example:</strong></p><pre><code class="language-julia hljs"># Supply the path to a .fastq file that you would like to convert to a FastRecord
myfasta = FastqtoFasta(&quot;myfastq.fastq&quot;);

# Alternatively, a FastqRecord can be used as the input 
myfasta = FastqtoFasta(myFastqRecord);</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/phorve/Nucleotide_Essentials.jl/blob/ef9c5233eb534080761fa3a827c431069ffb4fe6/src/Writing/FastqtoFasta.jl#L1-L25">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Nucleotide_Essentials.FilterQuality_se" href="#Nucleotide_Essentials.FilterQuality_se"><code>Nucleotide_Essentials.FilterQuality_se</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">Nucleotide_Essentials.FilterQuality_se</code></pre><p>Filters an input .fastq file based upon the encoded Phred+33 or Phred+64 quality scores. The encoding of the reads is automatically deteremined by looking for unique encoding in Phred+33 and Phred+64. Phred+64 encoding is identified by searching for <code>^</code>, <code>a</code>, <code>]</code>, and <code>f</code>.</p><p>Reads are filtered based upon the number of <a href="https://academic.oup.com/bioinformatics/article/31/21/3476/194979?login=true">expected errors</a> (<span>$\mathrm{E}$</span>) based on the error rate based on quality score and the sum of error probabilities, following the equation: </p><p><span>$\mathrm{E} = \sum{_ip_i} = \sum{_i}10^{\frac{-Q_i}{10}}$</span></p><p>Stringent filtering (<code>maxEE = 1</code>) is used by default but can be adjusted by the user. </p><p>Reads that pass the filtering parameters are output to a file ending in <code>_FilteredReads.fastq</code> in the user-determined directory, as indicated by <code>out</code>. </p><p>Supported keyword arguments include: </p><ul><li><code>read1::String</code>: Path to the reads to undergo quality filtering </li><li><code>out::String</code>: Path to the directory where reads that pass the quality filtering should be written </li><li><code>maxEE::Int64</code> (optional): The max number of expected errors a read can include as the filtering parameter (default: <code>maxEE = 1</code>)</li><li><code>verbose::Bool</code> (optional): Whether or not to show some intermediary feedback on the progress of the function (default = false)</li></ul><p><strong>Example:</strong></p><pre><code class="language-julia hljs">FilterQuality_se(&quot;forward_R1.fasta&quot;, &quot;/outdirectory&quot;)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/phorve/Nucleotide_Essentials.jl/blob/ef9c5233eb534080761fa3a827c431069ffb4fe6/src/QualityTesting/FilterQuality.jl#L3-L27">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Nucleotide_Essentials.FilterQuality_pe" href="#Nucleotide_Essentials.FilterQuality_pe"><code>Nucleotide_Essentials.FilterQuality_pe</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">Nucleotide_Essentials.FilterQuality_pe</code></pre><p>Filters an input .fastq file based upon the encoded Phred+33 or Phred+64 quality scores. The encoding of the reads is automatically deteremined by looking for unique encoding in Phred+33 and Phred+64. Phred+64 encoding is identified by searching for <code>^</code>, <code>a</code>, <code>]</code>, and <code>f</code>.</p><p>Reads are filtered based upon the number of <a href="https://academic.oup.com/bioinformatics/article/31/21/3476/194979?login=true">expected errors</a> (<span>$\mathrm{E}$</span>) based on the error rate based on quality score and the sum of error probabilities, following the equation: </p><p><span>$\mathrm{E} = \sum{_ip_i} = \sum{_i}10^{\frac{-Q_i}{10}}$</span></p><p>Stringent filtering (<code>maxEE = 1</code>) is used by default but can be adjusted by the user. </p><p>Output Files: </p><ul><li>If both the forward and reverse reads pass the filtering parameters: <ul><li>Forward reads are output to a file ending in <code>R1_Paired_filtered.fastq</code> in the user-determined directory, as indicated by <code>out</code></li><li>Reverse reads are output to a file ending in <code>R2_Paired_filtered.fastq</code> in the user-determined directory, as indicated by <code>out</code></li></ul></li><li>If only the forward read passes the filtering parameters: <ul><li>Forward reads are output to a file ending in <code>R1_Unpaired_filtered.fastq</code> in the user-determined directory, as indicated by <code>out</code></li><li>Reverse reads are not written to a file </li></ul></li><li>If only the reverse read passes the filtering parameters: <ul><li>Reverse reads are output to a file ending in <code>R2_Unpaired_filtered.fastq</code> in the user-determined directory, as indicated by <code>out</code></li><li>Forward reads are not written to a file </li></ul></li></ul><p>Supported keyword arguments include: </p><ul><li><code>read1::String</code>: Path to the forward reads to undergo quality filtering </li><li><code>read2::String</code>: Path to the reverse reads to undergo quality filtering </li><li><code>out::String</code>: Path to the directory where reads that pass the quality filtering should be written </li><li><code>maxEE::Int64</code> (optional): The max number of expected errors a read can include as the filtering parameter (default: <code>maxEE = 1</code>)</li><li><code>verbose::Bool</code> (optional): Whether or not to show some intermediary feedback on the progress of the function (default = false)</li></ul><p><strong>Example:</strong></p><pre><code class="language-julia hljs">FilterQuality_pe(&quot;forward_R1.fasta&quot;, &quot;reverse_R2.fasta&quot;, &quot;/outdirectory&quot;)

# changing the filtering parameters 
FilterQuality_pe(&quot;forward_R1.fasta&quot;, &quot;reverse_R2.fasta&quot;, &quot;/outdirectory&quot;, 2, true)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/phorve/Nucleotide_Essentials.jl/blob/ef9c5233eb534080761fa3a827c431069ffb4fe6/src/QualityTesting/FilterQuality.jl#L302-L341">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Nucleotide_Essentials.PlotQuality" href="#Nucleotide_Essentials.PlotQuality"><code>Nucleotide_Essentials.PlotQuality</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">Nucleotide_Essentials.PlotQuality</code></pre><p>Returns a plot of the quality profile of a .fastq or .fastq.gz file </p><p>This function plots a visual summary of the distribution of quality scores (automatically detects Phred+33 or Phred+64 encoding) as a function of sequence position for the input fastq file(s).</p><p>The plotted lines show summary statistics at each sequence position:</p><ul><li>green is the mean </li><li>dashed red lines are the 25th and 75th quantiles</li></ul><p>Supported keyword arguments include:</p><ul><li><code>Input::FastqRecord</code>: The name of a FastqRecord for plotting</li><li><code>verbose::Bool</code> (optional): Whether or not to show some intermediary feedback on the progress of the function (default = false)</li><li><code>outputfigure::Bool</code> (optional): Whether or not to output a .png file with the created QualityPlot (default = false)</li><li><code>figurepath::String</code> (optional): If outputting a .png figure to file, specify the path to a directory where the file should be written to (default = <code>pwd()</code>)</li></ul><p><strong>Example:</strong></p><pre><code class="language-julia hljs"># A quality profile can be created by supply the path to a .fastq or .fastq.gz file 
PlotQuality(&quot;path/to/my/file.fastq&quot;)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/phorve/Nucleotide_Essentials.jl/blob/ef9c5233eb534080761fa3a827c431069ffb4fe6/src/QualityTesting/PlotQuality.jl#L7-L31">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Nucleotide_Essentials.potential_mismatches" href="#Nucleotide_Essentials.potential_mismatches"><code>Nucleotide_Essentials.potential_mismatches</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">Nucleotide_Essentials.potential_mismatches</code></pre><p>Returns an <code>Vector{Any}</code> of potential barcodes with a single nucleotide change, including both deletions and substitutions</p><p>Supported keyword arguments include:</p><ul><li><code>Path::String</code>: The full or relative path to a .fastq file</li><li><code>mismatch::Int64</code>: The number of altered nucleotides to include (1 is only supported at this time)</li></ul><p><strong>Example:</strong></p><pre><code class="language-julia hljs">potential_mismatches(&quot;GCGT&quot;, 1)
17-element Vector{Any}:
&quot;GCGT&quot;
&quot;CCGT&quot; 
&quot;ACGT&quot; 
&quot;TCGT&quot; 
&quot;GGGT&quot; 
&quot;GAGT&quot; 
&quot;GTGT&quot; 
&quot;GCCT&quot; 
&quot;GCAT&quot; 
&quot;GCTT&quot; 
&quot;GCGG&quot; 
&quot;GCGC&quot; 
&quot;GCGA&quot; 
&quot;CGT&quot;
&quot;GGT&quot;
&quot;GCT&quot; 
&quot;GCG&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/phorve/Nucleotide_Essentials.jl/blob/ef9c5233eb534080761fa3a827c431069ffb4fe6/src/Demultiplex/Demultiplex.jl#L3-L35">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Nucleotide_Essentials.reverse_complement" href="#Nucleotide_Essentials.reverse_complement"><code>Nucleotide_Essentials.reverse_complement</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">Nucleotide_Essentials.reverse_complement</code></pre><p>Takes a string of nucleotide bases and returns the reverse complement of that string. Accepts inputs of <code>String</code> and <code>SubString{String}</code> (input from a FastqRecord)</p><p>Supported keyword arguments include:</p><ul><li><code>sequence::Union{String, SubString{String}}</code>: A string sequence of nucleotide bases or sequence entry from a FastqRecord</li></ul><p><strong>Example:</strong></p><pre><code class="language-julia hljs">reverse_complement(&quot;ATCGT&quot;)
&quot;ACGAT&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/phorve/Nucleotide_Essentials.jl/blob/ef9c5233eb534080761fa3a827c431069ffb4fe6/src/Demultiplex/Demultiplex.jl#L66-L80">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Nucleotide_Essentials.demultiplex_se" href="#Nucleotide_Essentials.demultiplex_se"><code>Nucleotide_Essentials.demultiplex_se</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">Nucleotide_Essentials.demultiplex_se</code></pre><p>Compares a list of provided barcodes with the provided multiplexed reads and separates the reads into individual .fastq files. If a barcode is found within the read, the barcode is removed from the sequence. The quality data of the reads is preserved and written to the outputted .fastq file. If a barcode is not found, the sequnce and quality is written to the unassigned .fastq file unchanged. </p><p>The mapping file must be either a .csv or .txt file with two columns. The first column heading must be <code>SampleID</code> and the second column heading must be <code>BarcodeSequence</code>. </p><p>EXAMPLE MAPPING FILE: </p><table><tr><th style="text-align: right">SampleID</th><th style="text-align: right">BarcodeSequence</th></tr><tr><td style="text-align: right">Sample1</td><td style="text-align: right">Barcode1</td></tr><tr><td style="text-align: right">Sample2</td><td style="text-align: right">Barcode2</td></tr><tr><td style="text-align: right">Sample3</td><td style="text-align: right">Barcode3</td></tr><tr><td style="text-align: right">Sample4</td><td style="text-align: right">Barcode4</td></tr><tr><td style="text-align: right">Sample5</td><td style="text-align: right">Barcode5</td></tr><tr><td style="text-align: right">Sample6</td><td style="text-align: right">Barcode6</td></tr><tr><td style="text-align: right">Sample7</td><td style="text-align: right">Barcode7</td></tr><tr><td style="text-align: right">Sample8</td><td style="text-align: right">Barcode8</td></tr></table><p>Supported keyword arguments include:</p><ul><li><code>R1::String</code>: Path to multiplexed reads   </li><li><code>Map::String</code>: Path to the mapping file </li><li><code>mismatch::Int64=0</code> (optional): Number of allowed mismatches in barcode. Potential options include 0 or 1. If 1 mismatch, computation time will significantly increase. Default is to allow for 0 mismatches (exact matches only). </li><li><code>debug::Bool=false</code> (optional): If true, a log file will be created and debugging data will be printed while the function is running (default is false).</li></ul><p><strong>Example:</strong></p><pre><code class="language-julia hljs">demultiplex_se(&quot;multiplexreads.fastq&quot;, &quot;mapping_file.fastq&quot;)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/phorve/Nucleotide_Essentials.jl/blob/ef9c5233eb534080761fa3a827c431069ffb4fe6/src/Demultiplex/Demultiplex.jl#L89-L120">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Nucleotide_Essentials.demultiplex_pe" href="#Nucleotide_Essentials.demultiplex_pe"><code>Nucleotide_Essentials.demultiplex_pe</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">Nucleotide_Essentials.demultiplex_pe</code></pre><p>Compares a list of provided barcodes with the provided paired-end multiplexed reads and separates the reads into individual .fastq files. If a barcode is found within R1 reads, the barcode is removed from the sequence. The quality data of the reads is preserved and written to the outputted .fastq file. If a barcode is not found, the sequnce and quality is written to the R1 unassigned .fastq file unchanged. If a barcode is found within R2 reads, the barcode is removed from the sequence. The quality data of the reads is preserved and written to the outputted .fastq file. If a barcode is not found, the sequnce and quality is written to the R2 unassigned .fastq file unchanged. </p><p><em>Dual-indexed reads are not <em>yet</em> supported</em></p><p>The mapping file must be either a .csv or .txt file with two columns. The first column heading must be <code>SampleID</code> and the second column heading must be <code>BarcodeSequence</code>. </p><p>EXAMPLE MAPPING FILE: </p><table><tr><th style="text-align: right">SampleID</th><th style="text-align: right">BarcodeSequence</th></tr><tr><td style="text-align: right">Sample1</td><td style="text-align: right">Barcode1</td></tr><tr><td style="text-align: right">Sample2</td><td style="text-align: right">Barcode2</td></tr><tr><td style="text-align: right">Sample3</td><td style="text-align: right">Barcode3</td></tr><tr><td style="text-align: right">Sample4</td><td style="text-align: right">Barcode4</td></tr><tr><td style="text-align: right">Sample5</td><td style="text-align: right">Barcode5</td></tr><tr><td style="text-align: right">Sample6</td><td style="text-align: right">Barcode6</td></tr><tr><td style="text-align: right">Sample7</td><td style="text-align: right">Barcode7</td></tr><tr><td style="text-align: right">Sample8</td><td style="text-align: right">Barcode8</td></tr></table><p>Supported keyword arguments include:</p><ul><li><code>R1::String</code>: Path to forward multiplexed reads   </li><li><code>R2::String</code>: Path to reverse multiplexed reads   </li><li><code>Map::String</code>: Path to the mapping file </li><li><code>mismatch::Int64=0</code> (optional): Number of allowed mismatches in barcode. Potential options include 0 or 1. If 1 mismatch, computation time will significantly increase. Default is to allow for 0 mismatches (exact matches only). </li><li><code>debug::Bool=false</code> (optional): If true, a log file will be created and debugging data will be printed while the function is running (default is false).</li></ul><p><strong>Example:</strong></p><pre><code class="language-julia hljs">demultiplex_pe(&quot;forward_multiplexreads.fastq&quot;, &quot;reverse_multiplexreads.fastq&quot;, &quot;mapping_file.fastq&quot;)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/phorve/Nucleotide_Essentials.jl/blob/ef9c5233eb534080761fa3a827c431069ffb4fe6/src/Demultiplex/Demultiplex.jl#L343-L377">source</a></section></article><h2 id="Index"><a class="docs-heading-anchor" href="#Index">Index</a><a id="Index-1"></a><a class="docs-heading-anchor-permalink" href="#Index" title="Permalink"></a></h2><ul><li><a href="#Nucleotide_Essentials.FastaRecord"><code>Nucleotide_Essentials.FastaRecord</code></a></li><li><a href="#Nucleotide_Essentials.FastqRecord"><code>Nucleotide_Essentials.FastqRecord</code></a></li><li><a href="#Nucleotide_Essentials.FastqtoFasta"><code>Nucleotide_Essentials.FastqtoFasta</code></a></li><li><a href="#Nucleotide_Essentials.FilterQuality_pe"><code>Nucleotide_Essentials.FilterQuality_pe</code></a></li><li><a href="#Nucleotide_Essentials.FilterQuality_se"><code>Nucleotide_Essentials.FilterQuality_se</code></a></li><li><a href="#Nucleotide_Essentials.PlotQuality"><code>Nucleotide_Essentials.PlotQuality</code></a></li><li><a href="#Nucleotide_Essentials.demultiplex_pe"><code>Nucleotide_Essentials.demultiplex_pe</code></a></li><li><a href="#Nucleotide_Essentials.demultiplex_se"><code>Nucleotide_Essentials.demultiplex_se</code></a></li><li><a href="#Nucleotide_Essentials.potential_mismatches"><code>Nucleotide_Essentials.potential_mismatches</code></a></li><li><a href="#Nucleotide_Essentials.readFasta"><code>Nucleotide_Essentials.readFasta</code></a></li><li><a href="#Nucleotide_Essentials.readFastq"><code>Nucleotide_Essentials.readFastq</code></a></li><li><a href="#Nucleotide_Essentials.reverse_complement"><code>Nucleotide_Essentials.reverse_complement</code></a></li><li><a href="#Nucleotide_Essentials.writeFasta"><code>Nucleotide_Essentials.writeFasta</code></a></li></ul><h2 id="Change-Log"><a class="docs-heading-anchor" href="#Change-Log">Change Log</a><a id="Change-Log-1"></a><a class="docs-heading-anchor-permalink" href="#Change-Log" title="Permalink"></a></h2><h4 id="Nucleotide_Essentials-v0.2.0"><a class="docs-heading-anchor" href="#Nucleotide_Essentials-v0.2.0">Nucleotide_Essentials v0.2.0</a><a id="Nucleotide_Essentials-v0.2.0-1"></a><a class="docs-heading-anchor-permalink" href="#Nucleotide_Essentials-v0.2.0" title="Permalink"></a></h4><ul><li>Added support for quality filtering of .fastq reads </li><li>Added support for Gzip compressed files </li><li>Performance improvements in <code>PlotQuality()</code> and added support for exporting quality plots</li><li>Added support for automatic quality profile encoding detection (Phred+64 and Phred+33 encoding)</li><li>Minor documentation updates </li></ul></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.15 on <span class="colophon-date" title="Thursday 14 April 2022 04:26">Thursday 14 April 2022</span>. Using Julia version 1.7.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
